contributors/design-proposals/storage/volume-reconstruct.md

```diff
@@ -0,0 +1,88 @@
+
+## Background and Motivation
+
+Today, kubelet volume manager uses desired/actual states to store some volume
+and pod information for reconciler to work upon. When kubelet restarts, all
+states become empty. Desired state populator will recover the desired states by
+getting the existing pods information from pod manager. Reconciler can then
+recover actual state through reconcile loop. 
+
+The problem is that if pod is deleted during kubelet restarts, desired state
+no longer has this pod and volume information so as the actual state. But the
+actual mount points and direcotries are still exist on the disk. In order to
+clean up those left over mounts and directories, we propose a volume
+reconstruction process. The basic idea is to scan pod directories and recover
+volume information (such as volume spec) from the mount path and device mount
+path which will be updated into the actual state. With this reconstructure
+process, actual state can match the real world and reconciler will tear down
+volumes accordingly. 
+
+However, this reconstruction design has issues for some plugins such as ()
+because there is not enough information to reconstruct volume spec correctly.
+For those volume plugins, volume tear down will fail.
+
+## Possible solution
+
+In searching of the solutions, we first need to understand why we need to reconstruct volume spec. 
+Volume spec is an internal representation of a volume.  All API volume types translate to Spec.
+```
+type Spec struct {
+	Volume           *v1.Volume                // for inline volume use case
+	PersistentVolume *v1.PersistentVolume      // for pvc use case
+	ReadOnly         bool
+}
+```
+Both Volume and PersistentVolume has a name and volume source. Different volume plugins have different representation of their sources. 
+Volume spec is constructed when adding pod volume information into desired state.
+The main use of this spec is to get the volume source name to construct unique volume name (which typicall includes plugin name and volume source name). 
+The unique volume name is used as a key in actual/desired state. The volume source name is also used on construct global mount path (device mount path)
+for volume mounter. When volume is teared down, unmounter uses the same function to construct the mount path from the spec again.
+Different from global mount path, the bind mount path uses plugin name and spec name (not the volume source name). 
+
+So look more closely, you can see for only the purpose of volume tear down, if we have a record of those mount paths (bind and global) in actual state, there is no need to 
+construct those paths from the volume spec again. In case of kubelet restarts, those mount paths could be scanned from the mount lists left on the disk. (bind mount path has
+a path structure which can be easily idenetified and global mount path could be obtained from the mount reference of the bind mount path). This observation leads to our first
```

Comment:
Should we consider the scenario that bind mount path not exist but global mount path still remains? Or we just solve this problem in another PR?

Reply:
In this case, constructVolumeSpec should fail, then in this proposal reconcile will try to tear down volume based on the mount path. Tear down process will remove the directory, but the global mount still exist. We can get the global mount path through the mount references from the bind mount path. If bind mount path does not exist, we need to figure out another way.